banditpylib/learners/collaborative_learner/utils.py

```diff
@@ -0,0 +1,185 @@
+
+  :param Optional[str] name: alias name
+  """
+
+  def __init__(self, name: Optional[str]):
+    self.__name = self._name() if name is None else name
+
+  @property
+  def name(self) -> str:
+    """Name of the agent"""
+    return self.__name
+
+  @abstractmethod
+  def _name(self) -> str:
+    """
+    Returns:
+      default agent name
+    """
+
+  @abstractmethod
+  def reset(self):
+    """Reset the agent
+
+    .. warning::
+      This function should be called before the start of each game.
+    """
+
+  @abstractmethod
+  def set_input_arms(self, arms: List[int]):
+    """Assign a set of arms to the agent
+
+    Args:
+      arms: arm indices that have been assigneds
+    """
+
+  @abstractmethod
+  def actions(self, context: Context) -> Actions:
+    """Actions of the agent
+
+    Args:
+      context: contextual information about the bandit environment
+
+    Returns:
+      actions to take
+    """
+
+  @abstractmethod
+  def update(self, feedback: Feedback):
+    """Update the agent
+
+    Args:
+      feedback: feedback returned by the bandit environment after
+        :func:`actions` is executed
+    """
+
+  @property
+  @abstractmethod
+  def best_arm(self) -> int:
+    """Arm that the agent chose (could be None)"""
+
+  @abstractmethod
+  def broadcast(self) -> Dict[int, Tuple[float, int]]:
+    """Broadcasts information learnt in the current round
+
+    Returns: (a dict of)
+      * arm ids used in learning
+      * Tuple[corresponding average reward seen,
+        number of pulls used to deduce average]
+    """
+
+
+class CollaborativeBAIMaster(ABC):
+  r"""Master that handles arm assignment and elimination
+
+  :param Optional[str] name: alias name
+  """
+
+  def __init__(self, name: Optional[str]):
+    self.__name = self._name() if name is None else name
+
+  @property
+  def name(self) -> str:
+    """Name of the master"""
+    return self.__name
+
+  @abstractmethod
+  def _name(self) -> str:
+    """
+    Returns:
+      default master name
+    """
+
+  @abstractmethod
+  def reset(self):
+    """Reset the master
+
+    .. warning::
+      This function should be called before the start of each game.
+    """
+
+  @abstractmethod
+  def initial_arm_assignment(self) -> Dict[int, List[int]]:
+    """The arm assignment for the first round
+
+    Returns:
+      dictionary of arm assignment per agent for all agents
+    """
+
+  @abstractmethod
+  def elimination(self, agent_ids: List[int],
+    messages: Dict[int, Tuple[float, int]]) -> Dict[int, List[int]]:
+    """Update the set of active arms based on some criteria
+    and return arm assignment
+
+    Args:
+      agent_ids: list of agents that will be assigned arms
+      messages: aggregation of messages broadcasted from agents
+
+    Returns:
+      dictionary of arm assignment per agent
+    """
+
+  @property
+  @abstractmethod
+  def active_arms(self):
+    """Arm indices that haven't been eliminated"""
```

Comment:
This method is not needed?

Reply:
Only needed if we get best arm from master

Reply:
See previous comment that we should ask agents for the final answer.